File: Java/src/main/java/Solution101_200/Solution141.java
package Solution101_200;

import tool.ListNode;

import java.util.HashMap;
import java.util.HashSet;

/**
 * @author windows
 */
public class Solution141 {
    public boolean hasCycle(ListNode head) {
        HashSet<ListNode> set = new HashSet<>();
        while(head != null) {
            if (set.contains(head)){
                return true;
            }else{
                set.add(head);
            }
            head = head.next;
        }
        return false;
    }


    public static void main(String[] args) {
        Solution141 solution141 = new Solution141();
    }
}

File: C++/src/Solution101_200/Solution141.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution141.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;



//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

bool hasCycle(ListNode *head) {
    if(head == nullptr || head->next == nullptr) return false;
    ListNode* slow = head;
    ListNode* fast = head->next;
    while(slow != fast){
        if(fast == nullptr || fast->next == nullptr) return false;
        slow = slow->next;
        fast = fast->next->next;
    }
    return true;
}